File: app/src/main/java/com/codekatana/passwordgen/MainActivity.java
package com.codekatana.passwordgen;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.SeekBar;
import android.widget.TextView;
import android.widget.Toast;

import java.util.List;
import java.util.Random;
import java.util.concurrent.ExecutionException;

public class MainActivity extends AppCompatActivity {


    public static final int MAX_RAND = 15;
    private SeekBar sb = null;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        sb = findViewById(R.id.seekNumWords);
        final TextView numWords = findViewById(R.id.lblTxtNumWords);
        sb.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
                numWords.setText(Integer.toString(i));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });
    }


    public void onClickGenerate(View v) {
        Random rand = new Random();
        AsyncTask<Integer, Integer, List<String>> task = new WikiDownloader().execute(sb.getProgress());

        EditText textBox = findViewById(R.id.txtGenerated);
        CheckBox chkUppercase = findViewById(R.id.chkUpperCase);
        CheckBox chkNumbers = findViewById(R.id.chkNumbers);
        CheckBox chkSymbols = findViewById(R.id.chkSymbols);

        boolean hasNumber = chkNumbers.isChecked();
        boolean hasSymbols = chkSymbols.isChecked();
        boolean hasUppercase = chkUppercase.isChecked();

        String generated = "";
        try {
            List<String> result = task.get();
            for (String word : result) {
                if (hasUppercase) {
                    generated = generated + word.substring(0, 1).toUpperCase() + word.substring(1);
                } else {
                    generated = generated + word;
                }
            }

            if (hasNumber) {
                generated = generated.replaceFirst("[oO]", "0");
                generated = generated.replaceFirst("[eE]", "3");
            }

            if (hasSymbols) {
                int powerBallNumber = rand.nextInt(MAX_RAND);
                if (powerBallNumber % 2 == 0)
                    generated = generated.replaceFirst("[sS]", "\\$");
                else
                    generated = generated.replaceFirst("[iI]", "!");
                if (powerBallNumber % 3 == 0)
                    generated = generated.replaceFirst("[aA]", "@");
            }

            textBox.setText(generated);
            TextView numChars = findViewById(R.id.lblNumChars);
            numChars.setText(getResources().getText(R.string.txt_NumChars) + String.format("%d", generated.length()));

        } catch (InterruptedException e) {
            e.printStackTrace();
        } catch (ExecutionException e) {
            e.printStackTrace();
        }
    }

    public void onClickCopy(View view) {
        EditText textBox = findViewById(R.id.txtGenerated);
        String generated = textBox.getText().toString();
        ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
        ClipData clip = ClipData.newPlainText("Generated Password", generated);
        clipboard.setPrimaryClip(clip);

        Context context = getApplicationContext();
        CharSequence text = getResources().getText(R.string.txt_copied);
        int duration = Toast.LENGTH_LONG;

        Toast.makeText(context, text, duration).show();
    }

}

File: app/src/main/java/com/codekatana/passwordgen/MainActivity.kt
package com.codekatana.passwordgen

import android.annotation.SuppressLint
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.*
import kotlinx.coroutines.*
import java.util.*

class MainActivity : AppCompatActivity(), CoroutineScope by CoroutineScope(Dispatchers.Default) {

    private fun updateUI(wc: Int): String {
        val rand = Random()
        val hasNumber = findViewById<CheckBox>(R.id.chkNumbers).isChecked
        val hasSymbols = findViewById<CheckBox>(R.id.chkSymbols).isChecked
        val hasUppercase = findViewById<CheckBox>(R.id.chkUpperCase).isChecked
        val words = EzPassApplication.wordBank.getWords(wc).toMutableList()
        launch { EzPassApplication.wordProcessor.getWords() }
        var generated = ""
        val nextRand = rand.nextInt(words.size)
        if (hasUppercase) {
            var word = words[nextRand]
            word = word.substring(0, 1).toUpperCase(Locale.ENGLISH) + word.substring(1)
            words[nextRand] = word
        }
        generated = words.joinToString(" ")

        if (hasNumber) {
            generated = WordProcessor.sprinkleNumbers(generated)
        }

        if (hasSymbols) {
            generated = WordProcessor.specialize(generated)
        }
        return generated
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val sb = findViewById<SeekBar>(R.id.seekNumWords)

        sb.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            @SuppressLint("SetTextI18n")
            override fun onProgressChanged(seekBar: SeekBar, i: Int, b: Boolean) {
                findViewById<TextView>(R.id.lblTxtNumWords).text = "${i + 2}"
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })

        launch {
            EzPassApplication.wordProcessor.getWords()
        }

    }

    fun onClickGenerate(view: View) {
        val numWordsNeeded = findViewById<SeekBar>(R.id.seekNumWords).progress + 2
        launch {
            val generated = async { updateUI(numWordsNeeded) }
            val textBox = findViewById<EditText>(R.id.txtGenerated)
            val numChars = findViewById<TextView>(R.id.lblNumChars)
            val generatedVal = generated.await()
            val length = generatedVal.filterNot { ' ' == it }.length
            runOnUiThread {
                textBox.setText(generatedVal)
                numChars.text = getString(R.string.txt_NumChars, length)
            }
        }
    }

    fun onClickCopy(view: View) {
        val textBox = findViewById<EditText>(R.id.txtGenerated)
        val generated = textBox.text.toString().filterNot { it == ' ' }
        val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        val clip = ClipData.newPlainText("Generated Password", generated)
        clipboard.primaryClip = clip

        val context = applicationContext
        val text = getString(R.string.txt_copied)
        val duration = Toast.LENGTH_LONG

        Toast.makeText(context, text, duration).show()
    }

    override fun onDestroy() {
        super.onDestroy()
        cancel()
    }

    companion object {
        const val MAX_RAND = 15
    }

}
